File: leetcode/js/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];
  const obj = {
    ")": "(",
    "}": "{",
    "]": "[",
  };

  for (const p of s) {
    if (p === "(" || p === "{" || p === "[") {
      stack.push(p);
      continue;
    }

    if (stack.length === 0 || obj[p] !== stack.pop()) {
      return false;
    }
  }

  return stack.length === 0;
};

File: leetcode/rust/valid-parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        use std::collections::VecDeque;

        let mut stack: VecDeque<char> = VecDeque::new();

        for p in s.chars() {
            if p == '{' || p == '(' || p == '[' {
                stack.push_back(p);
            } else {
                let Some(c) = stack.pop_back() else {
                    return false;
                };

                if (p == ')' && c != '(') || (p == '}' && c != '{') || (p == ']' && c != '[') {
                    return false;
                }
            }
        }

        stack.is_empty()
    }
}
